usr/man/man8/init.8
.TH init 8
.SH NAME
init \- process control initialization
.SH SYNTAX
.B /etc/init
.SH DESCRIPTION
The ULTRIX system invokes the
.PN init
command as the last step in the boot procedure.
The system 
normally then runs the automatic reboot sequence, as described in
reboot(8).
If 
.PN reboot 
succeeds, 
.PN init
begins multi-user operation.
If 
.PN reboot 
fails, 
.PN init
begins single-user operation by giving
the superuser a shell on the console.  
.PP
You can use the BOOT command 
so that parameters are passed
from the boot program to
.PN init
so that multi-user operation begins immediately. 
When the superuser terminates the single-user
shell (by pressing CTRL-D),
.PN init
runs the
.PN /etc/rc 
command file
without the reboot parameter.
This command file
performs housekeeping operations
such as removing temporary files,
mounting file systems, and starting
daemons.
For further information, see reboot(8).
.PP
In multi-user operation, 
.PN init
creates a process for each
terminal port where a user may log in.
To begin such operations, it reads the file /etc/ttys.
For further information, see ttys(5).
For each terminal that is marked ``on'' in the 
.PN ttys
file, 
.PN init
forks and invokes the command specified for the current line. 
The command is passed the name of the terminal as the
last argument. 
The other arguments (if any) are specified after the command
in the 
.PN ttys
file.
Usually, the command is getty(8), 
but it may be any command.
.PP
The 
.PN getty
command reads the user's name and invokes
.PN login
to log in the user and execute the shell.
.PP
Ultimately, the shell terminates
because of an end-of-file.  The end-of-file may be 
typed explicitly or generated as a result of 
hanging up on a terminal line.
The main path of
.PN init ,
which has been waiting
for such an event,
wakes up and removes the appropriate entry from the
file
.PN utmp ,
which records current users.  
.PN init 
then makes an entry in /usr/adm/wtmp,
which maintains a history
of logins and logouts.
The
.PN wtmp
entry is made only if a user logged in successfully on the line.
Then the appropriate terminal is reopened and
.PN getty
is
reinvoked.
.PP
.PN init
catches the
.PN hangup
signal (signal SIGHUP) and interprets it to mean that
the file /etc/ttys should be read again.
The shell process on each line which used to be active
in
.PN ttys
but is no longer there is terminated;
a new process is created for each added line;
lines unchanged in the file are undisturbed.
Thus it is possible to drop or add phone lines without
rebooting the system by changing the
.PN ttys
file and sending a
.PN hangup
signal to the
.PN init
process: use ``kill \-HUP 1.''
.PP
.PN init
terminates multi-user operations and resumes single-user mode
if it receives a terminate (TERM) signal.  That is, the
superuser types ``kill \-TERM 1''.
If there are processes outstanding which are deadlocked (due to
hardware or software failure),
.PN init
does not wait for them all to die (which might take forever), but
times out after 30 seconds and prints a warning message.
.PP
If 
.PN init
receives a terminal stop signal (the superuser types
``kill \-TSTP 1''),
.PN init
stops creating new processes
and lets the system slowly die away. 
A later hangup will resume full
multi-user operations, or a terminate will initiate a single user shell.
This feature is used by reboot(8) and halt(8).
.PP
If
.PN init
dies, the system will reboot itself
automatically.
If, at bootstrap time, the
.PN init
process cannot be located, the system will loop in user mode at location
0x13.
.SH DIAGNOSTICS
\fBWARNING: Something is hung (wont die); ps axl advised\fR.
.RS
The system is shutting down, and
.PN init
cannot kill a certain process. 
This usually when a process
cannot exit a device driver
due to a persistent device error condition.
.RE
.PP
\fBinit: \fI`command  tty'\fB failing, sleeping.\fR
.RS
.PN init
tried to spawn a new process 
(use the execve(2) system call)
for the \fIcommand\fR five times.  Each time, the \fIcommand\fR 
failed.  This may indicate that the \fIcommand\fR was invoked with
invalid arguments.  Check the /etc/ttys file for errors.
This error message is printed at syslog(3) level LOG_ERR.
.RE
.LP
\fBinit: exec failed: cmd=\fI  command reason\fR.
.RS
.PN init
tried to spawn a new process
using the execve(2)
system call for the \fIcommand\fR.  The 
.PN execve 
failed.
\fIreason\fR is the explanation produced by the perror(3) 
routine for why the 
.PN execve
failed.
This error message is printed at 
syslog(3) level LOG_ERR.
.RE
.SH FILES
.IP /etc/utmp 1i
Lists current system users
.IP /usr/adm/wtmp 1i
History of logins and logouts
.IP /etc/ttys 1i
.PN init
reads this file for a command to execute for
the terminal line
.IP /etc/rc 1i
Command file executed by
.PN init
.SH "SEE ALSO"
ttys(5), getty(8), rc(8), reboot(8)


etc/init.c

#ifndef lint
static char *sccsid = "@(#)init.c	1.14    ULTRIX  1/29/87";
#endif lint
/*
    init

    Last step in the boot procedure.  Creates a process for each
    terminal port in multi-user mode.

- - - - - - -
Modification History
~~~~~~~~~~~~~~~~~~~~
01	19-Mar-86, Greg Tarsa
	- Added back opser code that got lost in a shuffle.

02	04-Dec-86, Robin Lewis
	- Added code to support the login limit key that lives
	  in /upgrade.

03	15-Dec-86, Robin Lewis
	- Encoded the string /upgrade so it would be harder to find
	  the point in the system that read from the /upgrade file.

04	16-Dec-86, Jeff Fries
	- Added code to remove the /etc/opseractive file prior
	  to starting up opser.

05	22-Jan-87, Larry Cohen
	- Increase the maximum command size from 128 to 400

*/

/************************************************************************
 *									*
 *			Copyright (c) 1984 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/

/* static	char *sccsid = "@(#)init.c	1.10 (Berkeley) 3/19/86"; */

#include <signal.h>
#include <sys/types.h>
#include <utmp.h>
#include <setjmp.h>
#include <sys/reboot.h>
#include <errno.h>
#include <sys/file.h>
#include <limits.h>
#include <ttyent.h>
#include <syslog.h>
#include <sys/param.h>
#include <sys/quota.h>

#define	LINSIZ	sizeof(wtmp.ut_line)
#define	CMDSIZ	400	/* max string length for getty or window command*/
#define	TABSIZ	256
#define	ALL	p = &itab[0]; p < &itab[TABSIZ]; p++
#define	EVER	;;
#define SCPYN(a, b)	strncpy(a, b, sizeof(a))
#define SCMPN(a, b)	strncmp(a, b, sizeof(a))

char	shell[]	= "/bin/sh";
char	minus[]	= "-";
char	runc[]	= "/etc/rc";
char	utmp[]	= "/etc/utmp";
char	wtmpf[]	= "/usr/adm/wtmp";
char	ctty[]	= "/dev/console";

/*
  Opser information.

  opser is the name of the program that should be run
  when opseractivefile exists.
*/
char	opser[] = "/opr/opser";
char	opseractivefile[] = "/etc/opseractive";

struct utmp wtmp;
struct	tab
{
	char	line[LINSIZ];
	char	comn[CMDSIZ];
	char	xflag;
	int	pid;
	int	wpid;		/* window system pid for SIGHUP	*/
	char	wcmd[CMDSIZ];	/* command to start window system process */
	time_t	gettytime;
	int	gettycnt;
} itab[TABSIZ];

int	fi;
int	mergflag;
char	tty[20];
jmp_buf	sjbuf, shutpass;
time_t	time0;

int	reset();
int	idle();
char	*strcpy(), *strcat();
long	lseek();

#define	mask(s)	(1 << ((s)-1))
struct	sigvec rvec = { reset, mask(SIGHUP), 0 };
/*
struct	sigvec rvec = { reset, sigmask(SIGHUP), 0 };
*/

#ifdef vax
main()
{
	register int r11;		/* passed thru from boot */
#else
main(argc, argv)
	char **argv;
{
#endif
	int howto, oldhowto;
	int fd;
	int x=2;
	char n;
	int i;
	char buf[1024];
/* coded /upgrade in ascii plus a constant 0x64 */
static char fpdata[] = {
	0x93,
	0xd9,
	0xd4,
	0xcb,
	0xd6,
	0xc5,
	0xc8,
	0xc9,
	0x64,
	0x64,
};
char fpstring[256];
	for(i=0;i<9;i++)
		fpstring[i]=fpdata[i] - 0x64;
	if((fd=open(fpstring,O_RDONLY)) >= 0)
	{
		i=read(fd,buf,sizeof(buf));
		if ( i > 0)
		{
		n=buf[87];
		if(buf[87]==((~buf[88])-5) && (buf[87] == (~buf[87+n+9])))
			x=n;
		}
	}
	close(fd);
	quota(Q_SLIMIT,0,x,0);
	time0 = time(0);
#ifdef vax
	howto = r11;
#else
	if (argc > 1 && argv[1][0] == '-') {
		char *cp;

		howto = 0;
		cp = &argv[1][1];
		while (*cp) switch (*cp++) {
		case 'a':
			howto |= RB_ASKNAME;
			break;
		case 's':
			howto |= RB_SINGLE;
			break;
		}
	} else {
		howto = RB_SINGLE;
	}
#endif
/*
	openlog("init", LOG_CONS|LOG_ODELAY, 0);
*/
	sigvec(SIGTERM, &rvec, (struct sigvec *)0);
	signal(SIGTSTP, idle);
	signal(SIGSTOP, SIG_IGN);
	signal(SIGTTIN, SIG_IGN);
	signal(SIGTTOU, SIG_IGN);
	(void) setjmp(sjbuf);
	for (EVER) {
		oldhowto = howto;
		howto = RB_SINGLE;
		if (setjmp(shutpass) == 0)
			shutdown();
		if (oldhowto & RB_SINGLE)
			single();
		if (runcom(oldhowto) == 0) 
			continue;
		merge();
		multiple();
	}
}

int	shutreset();

shutdown()
{
	register i;
	register struct tab *p;

	close(creat(utmp, 0644));
	signal(SIGHUP, SIG_IGN);
	for (ALL) {
		term(p);
		p->line[0] = 0;
	}
	signal(SIGALRM, shutreset);
	alarm(30);
	for (i = 0; i < 5; i++)
		kill(-1, SIGKILL);
	while (wait((int *)0) != -1)
		;
	alarm(0);
	shutend();
}

char shutfailm[] = "WARNING: Something is hung (wont die); ps axl advised\n";

shutreset()
{
	int status;

	if (fork() == 0) {
		int ct = open(ctty, 1);
		write(ct, shutfailm, sizeof (shutfailm));
		sleep(5);
		exit(1);
	}
	sleep(5);
	shutend();
	longjmp(shutpass, 1);
}

shutend()
{
	register i, f;

	acct(0);
	signal(SIGALRM, SIG_DFL);
	for (i = 0; i < 10; i++)
		close(i);
	f = open(wtmpf, O_WRONLY|O_APPEND);
	if (f >= 0) {
		SCPYN(wtmp.ut_line, "~");
		SCPYN(wtmp.ut_name, "shutdown");
		SCPYN(wtmp.ut_host, "");
		time(&wtmp.ut_time);
		write(f, (char *)&wtmp, sizeof(wtmp));
		close(f);
	}
	return (1);
}

single()
{
	register pid;
	register xpid;
	extern	errno;

	do {
		pid = fork();
		if (pid == 0) {
			signal(SIGTERM, SIG_DFL);
			signal(SIGHUP, SIG_DFL);
			signal(SIGALRM, SIG_DFL);
			signal(SIGTSTP, SIG_IGN);
			(void) open(ctty, O_RDWR);
			dup2(0, 1);
			dup2(0, 2);

			/* If file exists restart opser, not shell */
			if (access(opseractivefile,0) == 0) {
				unlink(opseractivefile);
				sync();
				sync();
				execl(opser, (char *)0);
				exit(0);
			}
			else {
				execl(shell, minus, (char *)0);
				exit(0);
			}
		}
		while ((xpid = wait((int *)0)) != pid)
			if (xpid == -1 && errno == ECHILD)
				break;
	} while (xpid == -1);
}

runcom(oldhowto)
	int oldhowto;
{
	register pid, f;
	int status;

	pid = fork();
	if (pid == 0) {
		(void) open("/", O_RDONLY);
		dup2(0, 1);
		dup2(0, 2);
		if (oldhowto & RB_SINGLE)
			execl(shell, shell, runc, (char *)0);
		else
			execl(shell, shell, runc, "autoboot", (char *)0);
		exit(1);
	}
	while (wait(&status) != pid)
		;
	if (status)
		return (0);
	f = open(wtmpf, O_WRONLY|O_APPEND);
	if (f >= 0) {
		SCPYN(wtmp.ut_line, "~");
		SCPYN(wtmp.ut_name, "reboot");
		SCPYN(wtmp.ut_host, "");
		if (time0) {
			wtmp.ut_time = time0;
			time0 = 0;
		} else
			time(&wtmp.ut_time);
		write(f, (char *)&wtmp, sizeof(wtmp));
		close(f);
	}
	return (1);
}

struct	sigvec	mvec = { merge, mask(SIGTERM), 0 };
/*
struct	sigvec	mvec = { merge, sigmask(SIGTERM), 0 };
*/

/*
 * Multi-user.  Listen for users leaving, SIGHUP's
 * which indicate ttys has changed, and SIGTERM's which
 * are used to shutdown the system.
 */
multiple()
{
	register struct tab *p;
	register pid;

	sigvec(SIGHUP, &mvec, (struct sigvec *)0);
	for (EVER) {
		pid = wait((int *)0);
		if (pid == -1)
			return;
		for (ALL) {
			/* must restart window system BEFORE emulator */
			if (p->wpid == pid || p->wpid == -1)
				wstart(p);
			if (p->pid == pid || p->pid == -1) {
				/* disown the window system */
				if (p->wpid)
					kill(p->wpid, SIGHUP);
				rmut(p);
				dfork(p);
			}
		}
	}
}

/*
 * Merge current contents of ttys file
 * into in-core table of configured tty lines.
 * Entered as signal handler for SIGHUP.
 */
#define	FOUND	1
#define	CHANGE	2
#define WCHANGE 4

merge()
{
	register struct tab *p;
	register struct ttyent *t;

	for (ALL)
		p->xflag = 0;
	setttyent();
	while (t = getttyent()) {
		if ((t->ty_status & TTY_ON) == 0)
			continue;
		for (ALL) {
			if (SCMPN(p->line, t->ty_name))
				continue;
			p->xflag |= FOUND;
			if (SCMPN(p->comn, t->ty_getty)) {
				p->xflag |= CHANGE;
				SCPYN(p->comn, t->ty_getty);
			}
			if (SCMPN(p->wcmd, t->ty_window)) {
				p->xflag |= WCHANGE|CHANGE;
				SCPYN(p->wcmd, t->ty_window);
			}
			goto contin1;
		}

		for (ALL) {
			if (p->line[0] != 0)
				continue;
			SCPYN(p->line, t->ty_name);
			p->xflag |= FOUND|CHANGE;
			SCPYN(p->comn, t->ty_getty);
			if (strcmp(t->ty_window, "") != 0) {
				p->xflag |= WCHANGE;
				SCPYN(p->wcmd, t->ty_window);
			}
			goto contin1;
		}
	contin1:
		;
	}
	endttyent();
	for (ALL) {
		if ((p->xflag&FOUND) == 0) {
			term(p);
			p->line[0] = 0;
			wterm(p);
		}
		/* window system should be started first */
		if (p->xflag&WCHANGE) {
			wterm(p);
			wstart(p);
		}
		if (p->xflag&CHANGE) {
			term(p);
			dfork(p);
		}
	}
}

term(p)
	register struct tab *p;
{

	if (p->pid != 0) {
		rmut(p);
		kill(p->pid, SIGKILL);
	}
	p->pid = 0;
	/* send SIGHUP to get rid of connections */
	if (p->wpid > 0)
		kill(p->wpid, SIGHUP);
}

#include <sys/ioctl.h>

dfork(p)
	struct tab *p;
{
	register pid;
	time_t t;
	int dowait = 0;

	time(&t);
	p->gettycnt++;
	if ((t - p->gettytime) >= 60) {
		p->gettytime = t;
		p->gettycnt = 1;
	} else if (p->gettycnt >= 5) {
		dowait = 1;
		p->gettytime = t;
		p->gettycnt = 1;
	}
	pid = fork();
	if (pid == 0) {
		signal(SIGTERM, SIG_DFL);
		signal(SIGHUP, SIG_IGN);
		if (dowait) {
			syslog(LOG_ERR, "init: '%s %s' failing, sleeping", p->comn, p->line);
			closelog();
			sleep(30);
		}
		execit(p->comn, p->line);
		exit(0);
	}
	p->pid = pid;
}

/*
 * Remove utmp entry.
 */
rmut(p)
	register struct tab *p;
{
	register f;
	int found = 0;

	f = open(utmp, O_RDWR);
	if (f >= 0) {
		while (read(f, (char *)&wtmp, sizeof(wtmp)) == sizeof(wtmp)) {
			if (SCMPN(wtmp.ut_line, p->line) || wtmp.ut_name[0]==0)
				continue;
			lseek(f, -(long)sizeof(wtmp), 1);
			SCPYN(wtmp.ut_name, "");
			SCPYN(wtmp.ut_host, "");
			time(&wtmp.ut_time);
			write(f, (char *)&wtmp, sizeof(wtmp));
			found++;
		}
		close(f);
	}
	if (found) {
		f = open(wtmpf, O_WRONLY|O_APPEND);
		if (f >= 0) {
			SCPYN(wtmp.ut_line, p->line);
			SCPYN(wtmp.ut_name, "");
			SCPYN(wtmp.ut_host, "");
			time(&wtmp.ut_time);
			write(f, (char *)&wtmp, sizeof(wtmp));
			close(f);
		}
		/*
		 * After a proper login force reset
		 * of error detection code in dfork.
		 */
		p->gettytime = 0;
	}
}

reset()
{

	longjmp(sjbuf, 1);
}

jmp_buf	idlebuf;

idlehup()
{

	longjmp(idlebuf, 1);
}

idle()
{
	register struct tab *p;
	register pid;

	signal(SIGHUP, idlehup);
	for (EVER) {
		if (setjmp(idlebuf))
			return;
		pid = wait((int *) 0);
		if (pid == -1) {
			sigpause(0);
			continue;
		}
		for (ALL) {
			/* if window system dies, mark it for restart */
			if (p->wpid == pid)
				p->wpid = -1;
			if (p->pid == pid) {
				rmut(p);
				p->pid = -1;
			}
		}
	}
}

wterm(p)
	register struct tab *p;
{
	if (p->wpid != 0) {
		kill(p->wpid, SIGKILL);
	}
	p->wpid = 0;
}

wstart(p)
	register struct tab *p;
{
	int npid = fork();

	if (npid == 0) {
/*
		signal(SIGTERM, SIG_DFL);
		signal(SIGHUP,  SIG_DFL);
		signal(SIGALRM, SIG_DFL);
		signal(SIGTSTP, SIG_IGN);
*/
		execit(p->wcmd, p->line);
		exit(0);
	}
	p->wpid = npid;
}

#define NARGS	20	/* must be at lease 4 */
#define ARGLEN	512	/* total size for all the argument strings */

execit(s, arg)
	char *s;
	char *arg;	/* last argument on line */
{
	char *argv[NARGS], args[ARGLEN], *envp[1];
	register char *sp = s;
	register char *ap = args;
	register char c;
	register int i, j;
	char buf[ARGLEN+40];

	/*
	 * First we have to set up the argument vector.
	 * "prog arg1 arg2" maps to exec("prog", "-", "arg1", "arg2"). 
	 */
	for (i = 1; i < NARGS - 2; i++) {
		argv[i] = ap;
		for (EVER) {
			if ((c = *sp++) == '\0' || ap >= &args[ARGLEN-1]) {
				*ap = '\0';
				goto done;
			}
			if (c == ' ') {
				*ap++ = '\0';
				while (*sp == ' ')
					sp++;
				if (*sp == '\0')
					goto done;
				break;
			}
			*ap++ = c;
		}
	}
done:
	argv[0] = argv[1];
	argv[1] = "-";
	argv[i+1] = arg;
	argv[i+2] = 0;
	envp[0] = 0;
	sigsetmask(0);
	execve(argv[0], &argv[1], envp);
	sprintf(buf, "init: exec failed: cmd= %s, %s", argv[0], argv[1]);
	/* report failure of exec */
	for (j=1; j<i+1; j++) {
		strcat(buf, ", ");
		strcat(buf, argv[j+1]);
	}
	syslog(LOG_ERR, "%s: %m", buf);
	closelog();
	sleep(10);	/* prevent failures from eating machine */
}
